### PR TITLE
Private Chain Provisioning

### PR DESCRIPTION
**Pull request by pelle**
This documents our experimental Private Chain Provisioning flow.

Before building anything on it, please contact us.

### CODE REVIEW

File: flows/privatechain.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+- Ethereum compatible blockchain
+- Public facing JSON RPC endpoint
+- [IdentityManager](https://github.com/uport-project/uport-identity/blob/develop/contracts/IdentityManager.sol) contract deployed on private blockchain
+- [Fueling Server](../rest-apis/fuel-server.md) for fueling transactions on private chain
+
+## Provisioning Methodology
+
+1. Use [Selective Disclosure Flow](selectivedisclosure.md) to request primary identity address of user
+2. Lookup public key in response from selective disclosure requst
```
Comment (ajunge) [inline flows/privatechain.md]: We actually look for the "deviceKey" that controls the identity. You mean "public key" ==  "deviceKey" ?

File: flows/privatechain.plantuml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+group Private Chain Provisioning
+  ProvisioningServer -> ProvisioningServer : Selective Disclosure Request
+  ProvisioningServer -> Owner : Show QR code containing request
+  Owner -> UportMobile: Scans QR code
+  UportMobile -> Owner : Authorize Request?
+  Owner -> UportMobile: Allow/Disallow
+  UportMobile -> UportMobile: Sign Selective Disclosure Response
+  UportMobile -> ProvisioningServer: Selective Disclosure Response
+  ProvisioningServer -> RpcGateway : Create Account Transaction
```
Comment (ajunge) [inline flows/privatechain.plantuml]: Before this step the "ProvisioningServer" should lookup for the deviceKey. That step is not reflected here

File: flows/privatechain.md [hunk 3]
```diff
@@ -15,12 +15,10 @@ The following shows the basic flow:
 
 ## Provisioning Methodology
 
-1. Use [Selective Disclosure Flow](selectivedisclosure.md) to request primary identity address of user
-2. Lookup public key in response from selective disclosure requst
-3. Calculate Ethereum address for public key
-4. Create transaction on IdentityManager contract calling `createIdentity` with the above ethereum address as the `owner`
-5. Lookup the `identity` field from the `LogIdentityCreated` event in the receipt from above transaction
-6. Create a valid signed [Private Chain Provisioning Message](../messages/privatechain.md) and send it to the uPort mobile app
+1. Use [Selective Disclosure Flow](selectivedisclosure.md) to request primary identity address of user as well as a new ethereum address known as the devicekey
```
Comment (ajunge) [inline flows/privatechain.md]: I will rephrase this as: "Use [Selective Disclosure Flow](selectivedisclosure.md) to request primary identity address of user as well as the deviceKey to create a new identity on the private blockchain"

File: flows/privatechain.md [hunk 4]
```diff
@@ -15,12 +15,10 @@ The following shows the basic flow:
...
+2. Create transaction on IdentityManager contract calling `createIdentity` with the devicekey as the `owner`
```
Comment (ajunge) [inline flows/privatechain.md]: MetaTx or "normal" ?

File: messages/privatechain.md [hunk 5]
```diff
@@ -15,8 +15,9 @@ Name | Description | Required
 [`sub`](https://tools.ietf.org/html/rfc7519#section-4.1.1) | The [MNID](https://github.com/uport-project/mnid) encoding of the [private chain account](https://github.com/uport-project/uport-identity/blob/develop/contracts/Proxy.sol)| yes
 [`iat`](https://tools.ietf.org/html/rfc7519#section-4.1.6) | The time of issuance | yes
 [`exp`](https://tools.ietf.org/html/rfc7519#section-4.1.4) | Expiration time of Verification | no
+`dad`|Ethereum address of devicekey as passed to provisioning service as `nad` in the [Selective Disclosure Response](./shareresp.md) | yes
```
Comment (ajunge) [inline messages/privatechain.md]: is this the `nad` of the shareResp ? If so, shouldn't be the called 'nad' instead of 'dad' ? (BTW, why restrict to three letters?)

File: messages/privatechain.md [hunk 6]
```diff
@@ -15,8 +15,9 @@ Name | Description | Required
...
 `ctl`|Ethereum address of the [controller contract](https://github.com/uport-project/uport-identity/blob/develop/contracts/MetaIdentityManager.sol) used to control the account | yes
 `reg`|Ethereum address of the [Uport Registry](https://github.com/uport-project/uport-registry/blob/master/contracts/UportRegistry.sol) used on private chain | no
-`fct`|Url of [fueling service](../rest-apis/fuel-server.md) for providing gas on private network | yes
+`fct`|Url of [fueling service](../rest-apis/fuel-server.md) for providing gas on private network | no
+`rel`|Url of [fueling service](../rest-apis/fuel-server.md) for providing gas on private network | no
```
Comment (ajunge) [inline messages/privatechain.md]: Why two? Right now we have `fct` (the old sensui style faucet) and `rel` (for the new metatx faucet). Can we stick only to MetaTx ?

Comment (pelle) [reply messages/privatechain.md]: They are both optional. So use the one you would like

File: messages/privatechain.md (hunk 5 above)
Comment (pelle) [reply messages/privatechain.md]: JWT tends to have small param names. `dad` is short for `device address`, but I think it makes sense to change it in the shareResponse. There will be future types as well.

File: flows/privatechain.md (hunk 4 above)
Comment (pelle) [reply flows/privatechain.md]: MetaTx for now
